File: backtracking_subset_sum.py
def subset_sum_util(numbers, target, current_subset, index, current_sum):
    # Base Case: Target achieved
    if current_sum == target:
        print(f"  {{ " + " ".join(map(str, current_subset)) + " }}")
        return

    # Loop through the numbers array
    for i in range(index, len(numbers)):
        if current_sum + numbers[i] <= target:
            
            # Action: Include number
            current_subset.append(numbers[i])

            # Recurse
            subset_sum_util(numbers, target, current_subset, i + 1, current_sum + numbers[i])

            # Backtrack
            current_subset.pop()

def main():
    print("--- DSA Masterclass: Subset Sum Problem (Python) ---\n")
    
    numbers = [10, 7, 5, 18, 12, 20]
    target = 35
    
    print(f"Target Sum: {target}")
    print(f"Subsets that sum up to {target} are:")
    print("------------------------------------------")
    subset_sum_util(numbers, target, [], 0, 0)
    print("------------------------------------------")

if __name__ == "__main__":
    main()

File: backtracking_subset_sum.cpp
#include <iostream>
#include <vector>

using namespace std;

// Print the subset found
void printSubset(const vector<int>& subset) {
    cout << "{ ";
    for (int num : subset) {
        cout << num << " ";
    }
    cout << "}" << endl;
}

// Backtracking utility function
void subsetSumUtil(const vector<int>& set, int target, vector<int>& currentSubset, int index, int currentSum) {
    // Base Case: If current sum matches the target
    if (currentSum == target) {
        printSubset(currentSubset);
        return;
    }

    // Explore further elements
    for (int i = index; i < set.size(); i++) {
        // Optimization step: Skip if adding this element exceeds the target
        if (currentSum + set[i] <= target) {
            
            // Include the current element
            currentSubset.push_back(set[i]);

            // Recur for remaining elements
            subsetSumUtil(set, target, currentSubset, i + 1, currentSum + set[i]);

            // BACKTRACK: Remove the element to explore other subsets
            currentSubset.pop_back();
        }
    }
}

int main() {
    cout << "--- DSA Masterclass: Subset Sum Problem (C++) ---" << endl << endl;

    vector<int> set = {10, 7, 5, 18, 12, 20};
    int target = 35;
    vector<int> currentSubset;

    cout << "Target Sum: " << target << endl;
    cout << "Subsets that sum up to " << target << " are:" << endl;
    cout << "------------------------------------------" << endl;
    
    subsetSumUtil(set, target, currentSubset, 0, 0);
    
    cout << "------------------------------------------" << endl;

    return 0;
}